[IR] Add `dead_on_return` attribute

antoniofrighetto opened the pull request:
Introduce `dead_on_return` attribute, which is meant to be taken advantage by the frontend, and states that the memory pointed to by the argument is dead upon function return. As with `byval`, it is supposed to be used for passing aggregates by value. The difference lies in the ABI: `byval` implies that the pointer is explicitly passed as argument to the callee (during codegen the copy is emitted as per byval contract), whereas a `dead_on_return`-marked argument may imply that the copy already exists at the IR level, is located at a specific stack offset within the caller, and this memory will not be read further by the caller upon callee return – or otherwise poison, if read before being written.

Related issue: https://github.com/llvm/llvm-project/issues/141200.

dtcxzyw commented on llvm/include/llvm/IR/Argument.h:
```diff
@@ -78,6 +78,9 @@ class Argument final : public Value {
   LLVM_ABI bool hasByValAttr() const;
 
+  /// Return true if this argument has the dead_on_return attribute.
+  bool hasDeadOnReturnAttr() const;
```
```suggestion
  LLVM_ABI bool hasDeadOnReturnAttr() const;
```

dtcxzyw commented on llvm/lib/Analysis/AliasAnalysis.cpp:
```diff
@@ -818,7 +818,7 @@ bool llvm::isNoAliasCall(const Value *V) {
 static bool isNoAliasOrByValArgument(const Value *V) {
   if (const Argument *A = dyn_cast<Argument>(V))
-    return A->hasNoAliasAttr() || A->hasByValAttr();
+    return A->hasNoAliasAttr() || A->hasByValAttr() || A->hasDeadOnReturnAttr();
```
Doesn't it imply that stores that would only be visible on the unwind path can also be elided?
If so, I don't think `dead_on_return` is a good name. It should be something that implies `dead_on_unwind`.

antoniofrighetto replied:
I think we may want to keep dead_on_return orthogonal from dead_on_unwind, and let the memory pointed by be dead upon normal function return. Landing pads may need to inspect parameter memory (not sure if can be the case?). Even so, that would otherwise subsume dead_on_unwind, and I don't think we want this?

nikic replied:
We also shouldn't check it here, this will imply aliasing properties.

nikic commented on llvm/docs/LangRef.rst:
```diff
@@ -1741,6 +1741,20 @@ Currently, only the following parameter attributes are defined:
 
     This attribute cannot be applied to return values.
 
+``dead_on_return``
```
The wording here should be a lot more similar to dead_on_unwind, describing what the precise operational semantics of the attribute are.

antoniofrighetto replied:
Tried to improve the wording, hope it aligns more now. PR description updated, too (I got slightly threw off by the extra temporary allocation done in the callee in the unoptimized IR when targeting AArch64).

nikic commented on llvm/docs/LangRef.rst:
```diff
@@ -1741,6 +1741,20 @@ Currently, only the following parameter attributes are defined:
+    *callee* explicitly allocates the temporary copy. Stores that would only be
+    visible on the normal return path may be optimized out. Likewise,
+    optimizations may assume that the pointer does not alias any memory that
+    outlives the call.
```
I'm not sure I understand what this last sentence means, but I don't think dead_on_return should have effects on aliasing. I'd expect that typically dead_on_return would be combined with the noalias attribute.

antoniofrighetto commented on llvm/lib/IR/Function.cpp:
```diff
@@ -176,7 +182,8 @@ bool Argument::hasPassPointeeByValueCopyAttr() const {
          Attrs.hasParamAttr(getArgNo(), Attribute::InAlloca) ||
-         Attrs.hasParamAttr(getArgNo(), Attribute::Preallocated);
+         Attrs.hasParamAttr(getArgNo(), Attribute::Preallocated) ||
+         Attrs.hasParamAttr(getArgNo(), Attribute::DeadOnReturn);
```
Oh, sure, updated.

antoniofrighetto commented on llvm/docs/LangRef.rst:
```diff
@@ -1741,6 +1741,23 @@ Currently, only the following parameter attributes are defined:
+    This attribute indicates that the memory pointed to by the argument is dead
+    upon normal function return, meaning that the caller will not depend on its
+    contents. Stores that would only be observable on the normal return path may
+    be elided.
```
Leaving this as is. Might be suboptimal (as we don't imply dead_on_unwind), but I also do expect to handle dead_on_return and dead_on_unwind in conjunction. Can revisit this if you wish (I'd also rather change name, in this case).

nikic reviewed (commented):
This looks good to me, but this should probably have a small RFC on discourse.

antoniofrighetto commented:
A small RFC can be found at: https://discourse.llvm.org/t/rfc-add-dead-on-return-attribute/86871.

nikic commented on llvm/test/Transforms/DeadStoreElimination/simple.ll:
Can you please also add a test variant that returns %p? To show that we can still elide the store in that case as well.

nikic reviewed (approved):
The RFC didn't generate any additional discussion, so LGTM.

antoniofrighetto committed:
[IR] Introduce `dead_on_return` attribute

Add `dead_on_return` attribute, which is meant to be taken advantage
by the frontend, and states that the memory pointed to by the argument
is dead upon function return. As with `byval`, it is supposed to be
used for passing aggregates by value. The difference lies in the ABI:
`byval` implies that the pointer is explicitly passed as argument to
the callee (during codegen the copy is emitted as per byval contract),
whereas a `dead_on_return`-marked argument implies that the copy
already exists in the IR, is located at a specific stack offset within
the caller, and this memory will not be read further by the caller upon
callee return – or otherwise poison, if read before being written.

RFC: https://discourse.llvm.org/t/rfc-add-dead-on-return-attribute/86871.
